### PR TITLE
DDF-2036 Drape image overlays on map (improved)

### PR DESCRIPTION
**Pull request by spearskw**
#### What does this PR do?
Improves the image overlays to look more accurate on the map

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@andrewkfiedler @jrnorth 

#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
[UI](https://github.com/orgs/codice/teams/ui)

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@jlcsmith
@kcwire

#### How should this be tested? (List steps with links to updated documentation)
Ingest some NITFs and click the "show overlay" button in the search UI. Check to make sure they look right.

#### Checklist:
- [x] Update / Add Unit Tests

### CODE REVIEW

File: catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java [hunk 1]
```diff
@@ -71,29 +83,106 @@ private BinaryContent overlay(Metacard metacard, Map<String, Serializable> argum
...
+
+        return createImage(image, boundary, (int) boundingBox.get(1).get(0), (int) -boundingBox.get(1).get(1));
+    }
+
+    private BinaryContent createBinaryContent(BufferedImage image)
+            throws CatalogTransformerException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
         try {
-            imageBuilder.toOutputStream(baos);
+            ImageIO.write(image, "png", baos);
```
Comment (bdeining) [inline catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java]: nit : you can use the PNG constant here

File: catalog/transformer/catalog-transformer-overlay/pom.xml [hunk 2]
```diff
@@ -26,6 +26,10 @@
     <name>DDF :: Catalog :: Transformer :: Overlay</name>
     <packaging>bundle</packaging>
 
+    <properties>
+        <la4j.version>0.6.0</la4j.version>
```
Comment (kcwire) [inline catalog/transformer/catalog-transformer-overlay/pom.xml]: You don't need to define a property for this if it's only used once within the same pom.

File: catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java [hunk 3]
```diff
@@ -71,29 +83,106 @@ private BinaryContent overlay(Metacard metacard, Map<String, Serializable> argum
...
+    private List<Vector> parseBoundary(String location) throws CatalogTransformerException {
+        final Geometry geometry = parseGeometry(location);
+        if (!canHandleGeometry(geometry)) {
+            throw new CatalogTransformerException("The metacard geometry is not a rectangle.");
+        }
+
+        final Coordinate[] coordinates = geometry.getCoordinates();
+
+        List<Vector> boundary = new ArrayList<>();
+        boundary.add(new BasicVector(new double[] {coordinates[0].x, coordinates[0].y}));
```
Comment (kcwire) [inline catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java]: Looping over the array would reduce the duplicate lines of code.

File: catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java [hunk 4]
```diff
@@ -71,29 +83,106 @@ private BinaryContent overlay(Metacard metacard, Map<String, Serializable> argum
...
     }
 
-    private BinaryContent createBinaryContent(Builder imageBuilder)
-            throws CatalogTransformerException {
-        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    private BufferedImage createTileFromImageAndBoundary(BufferedImage image, List<Vector> boundary) {
+        /*
+         * We transform the image by moving the corners and applying
+         * transparency so that it looks right when laid down as a north-up
+         * rectangular tile (the only thing Cesium currently supports).
```
Comment (kcwire) [inline catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java]: Remove Cesium from the comment - this should be map agnostic.

File: catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java [hunk 5]
```diff
@@ -71,29 +83,106 @@ private BinaryContent overlay(Metacard metacard, Map<String, Serializable> argum
...
+        return createBinaryContent(tile);
     }
 
-    private void rotate(Builder imageBuilder, String wkt) throws CatalogTransformerException {
-        final double angle = GeometryUtils.getRotationAngle(wkt);
-        imageBuilder.rotate(angle);
+    private List<Vector> parseBoundary(String location) throws CatalogTransformerException {
+        final Geometry geometry = parseGeometry(location);
+        if (!canHandleGeometry(geometry)) {
+            throw new CatalogTransformerException("The metacard geometry is not a rectangle.");
```
Comment (kcwire) [inline catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java]: Can we update this to `The Image boundary is not a rectangle` or something that would be more meaningful to a user?

File: catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java [hunk 6]
```diff
@@ -71,29 +83,106 @@ private BinaryContent overlay(Metacard metacard, Map<String, Serializable> argum
...
+        double scaleFactor = calculateScaleFactor(boundary, image.getWidth());
+        boundary.replaceAll(v -> v.multiply(scaleFactor));
+        boundingBox.replaceAll(v -> v.multiply(scaleFactor));
+
+        return createImage(image, boundary, (int) boundingBox.get(1).get(0), (int) -boundingBox.get(1).get(1));
+    }
+
+    private BinaryContent createBinaryContent(BufferedImage image)
+            throws CatalogTransformerException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
Comment (kcwire) [inline catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java]: Make this a try-with-resources to ensure the BAOS is closed when you're done with it.

File: catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java (hunk 4 above)
Comment (spearskw) [inline catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java]: I will remove that part of the comment. Other maps have the ability to put a picture in a boundary you define with a single function call, so I wanted to be reminded that if Cesium adds this functionality or if we switch to something else then we should revisit this.

File: catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java (hunk 5 above)
Comment (spearskw) [inline catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java]: Any suggestions for what might make more sense? The best I can come up with is "4 sided input implies 4 sided overlay", but that glosses over the fact that mosaics with more than 4 sides are treated as 4 sided.

**Comment by jlcsmith** [comment]
Hero'd successfully.  Tested with about a dozen different NITF images and they all looked good.  The accuracy of the overlays are improved considerably.

File: catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java (hunk 5 above)
Comment (spearskw) [inline catalog/transformer/catalog-transformer-overlay/src/main/java/ddf/catalog/transformer/OverlayMetacardTransformer.java]: Whoops misread that comment. Fixed.

**Pull request merged**